### ISSUE TITLE
using `@MappedSuperclass` can not native compile with Kotlin `val`

### ISSUE THREAD

**Issue by OrdinarySF** [quarkusio/quarkus#36251]
### Describe the bug

```
Fatal error: com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing org.acme.MyKotlinEntity.$$_hibernate_read_deletedTime()
Parsing context:
   at org.acme.MyKotlinEntity.$$_hibernate_read_deletedTime(MyKotlinEntity.kt)
   at org.acme.DeletableEntity.$$_hibernate_read_deletedTime(DeletableEntity.kt)

        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:153)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:104)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureFlowsGraphCreated(MethodTypeFlow.java:83)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.getOrCreateMethodFlowsGraph(MethodTypeFlow.java:65)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultVirtualInvokeTypeFlow.java:109)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:562)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.PointsToAnalysis$1.run(PointsToAnalysis.java:488)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:193)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:177)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
------------------------------------------------------------------------------------------------------------------------
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: org.acme.DeletableEntity.$$_hibernate_write_deletedTime(java.time.LocalDateTime). This error is reported at image build time because class org.acme.MyKotlinEntity is registered for linking at image build time by command line
        at parsing org.acme.MyKotlinEntity.$$_hibernate_read_deletedTime(MyKotlinEntity.kt)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.throwParserError(BytecodeParser.java:2518)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.throwParserError(SharedGraphBuilderPhase.java:110)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3393)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.handleBytecodeBlock(BytecodeParser.java:3345)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3190)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1138)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1030)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:97)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:84)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:446)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:135)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:685)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:171)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:349)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:93)
        ... 13 more
Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: org.acme.DeletableEntity.$$_hibernate_write_deletedTime(java.time.LocalDateTime). This error is reported at image build time because class org.acme.MyKotlinEntity is registered for linking at image build time by command line
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:333)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedMethod(SharedGraphBuilderPhase.java:323)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedInvoke(SharedGraphBuilderPhase.java:279)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeSpecial(BytecodeParser.java:1778)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeSpecial(BytecodeParser.java:1768)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5287)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3385)
        ... 28 more
```

### How to Reproduce?

https://github.com/OrdinarySF/quarkus-mappedsuperclass-demo

### Output of `uname -a` or `ver`

windows 11

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 LTS OpenJDK Runtime Environment GraalVM 22.3.2 (build 17.0.7+7-LTS) OpenJDK 64-Bit Server VM GraalVM 22.3.2 (build 17.0.7+7-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.4.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

------------------------------------------------------------ Gradle 8.2 ------------------------------------------------------------  Build time:   2023-06-30 18:02:30 UTC Revision:     5f4a070a62a31a17438ac998c2b849f4f6892877  Kotlin:       1.8.20 Groovy:       3.0.17 Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023 JVM:          17.0.7 (BellSoft 17.0.7+7-LTS) OS:           Windows 11 10.0 amd64

**Comment by gsmet** [comment]
@yrodiere looks like an issue with bytecode recording and reflection in the Hibernate ORM extension?

From what I can see, the plugin and BOM version are consistent so I suppose it's not a version mismatch.

**Comment by yrodiere** [comment]
@gsmet This could very well be a problem in ORM itself... there's been multiple problems with this "generate a setter call in getters" part, recently.

@marko-bekhta do you think you could have a look when you have some spare cycles? In particular we'd need to know whether this can be reproduced without Kotlin and without Quarkus, and if so we'd need to report it upstream. There are already a few open issues related to Kotlin in Hibernate ORM, so if it's specific to Kotlin maybe it's a known issue...
If it's specific to Quarkus feel free to ping me so that we debug this together.

**Comment by marko-bekhta** [comment]
Hey, sure. I'll take a look into this after the v5 helper tests.

**Comment by marko-bekhta** [comment]
A preliminary update:

The problem comes from using `val` in a kotlin mapped superclass:
```kotlin
@MappedSuperclass
abstract class DeletableEntity : BaseEntity() {
    @Column(nullable = true, insertable = false)
    protected val deletedTime: LocalDateTime? = null
}
```
^ changing from `val` to `var` will pass the native compilation. I've also managed to reproduce the problem with Java entities:
```java
@MappedSuperclass
public class MappedBase {
    @Column
    // Important to make the field private so that `FieldReaderAppender#of()` will go into the "else" branch.
    private final LocalDateTime updated = LocalDateTime.now();
}
@Entity
public class MyEntity extends MappedBase {
    @Id
    Long id;
}
```
From what it looks so far -- the problem comes from the ORM side. I'll look closer into it tomorrow and see if I can suggest a fix. In the meantime, the workaround for the issue could be changing from `val` to `var` even though that changes the field behaviour a bit...

**Comment by yrodiere** [comment]
This should get fixed in Quarkus when we upgrade to Hibernate ORM 6.4.1.

**Comment by yrodiere** [comment]
This was fixed by #38162
